### DISCUSSION TITLE
.bashrc does not seem to be loaded

### DISCUSSION THREAD

**Discussion by ssbarnea** [Vampire/setup-wsl#54] [Q&A] [answered]
Apparently the `.bashrc` file is not loaded regardless of setting the `bash -l` as login command.

```
      - name: Setup asdf
        if: ${{ !contains(matrix.shell, 'wsl') }}
        uses: asdf-vm/actions/install@v3

      - name: Install asdf inside WSL
        if: contains(matrix.shell, 'wsl')
        run: |
          set -ex
          git clone https://github.com/asdf-vm/asdf.git ~/.asdf --branch v0.14.0
          echo '. "$HOME/.asdf/asdf.sh"' >> ~/.bashrc
          export ASDF_DIR="$HOME/.asdf"
          . "$HOME/.asdf/asdf.sh"
          asdf plugin add task
          asdf install
          task --version

  	  - name: Task failing due to 'command not found'
  	    shell: "wsl-bash {0}"
        run: task --version
```

You can see the last task failing with command not found at https://github.com/ssbarnea/x0/actions/runs/8523493674/job/23345891540?pr=1#step:5:1

What is quite interesting is that when playing with WSL on my own host, this worked (installed asdf using commands in the first snipped and tested that another invocation of wsl would have asdf available.
```
wsl --install Ubuntu-22.04 --no-launch
ubuntu2204.exe --install --root
wsl --set-default Ubuntu-22.04
wsl --set-version Ubuntu-22.04 1
wsl --list -v
wsl --exec lsb_release -r
```

**Comment by Vampire** [comment]
That's correct and has not much to do with this action.
`.bashrc` is executed when you start an interactive shell.
`.bash_profile` is executed when you start a login shell.

**Reply by ssbarnea** [reply]
I did read about this on https://www.cloudwithchris.com/blog/fix-bashrc-not-executing-wsl/ but in absence of any `.bash_profile` file this should be loaded. 

https://shreevatsa.wordpress.com/2008/03/30/zshbash-startup-files-loading-order-bashrc-zshrc-etc/ is another useful resource. Based on it, it conclude that on a clean Ubuntu, `~/.bashrc` should be loaded due to the inclusion of it from `~/.profile` and the missing `~/.bash_profile`. 

![](https://shreevatsa.files.wordpress.com/2008/03/bashstartupfiles1.png)

Based on that graph, I am still inclined to believe that altering `~/.bashrc` file is the right approach, but I guess that I missed configuring the rights parameters for `wsl-shell-command` in order to ensure it loads it?

I wonder where I should install the asdf enablement code to ensure it is active for all types of shells.

**Reply by ssbarnea** [reply]
Kinda sorted by adding `wsl-shell-command: "bash -i"` - any other attempts did not make it load the `.bashrc` file.

I am bit surprised because based on the graph `--login` should also have being enough. In fact it would have being prefered because running on CI is not really "interactive", and setting this might have some undesired surprises (prompts?) later, I guess.

**Reply by Vampire** [reply]
Yeah, you were right, it works perfectly fine with both, interactive shell (which seems wrong to be used) as well as login shell. Have a look at this quick test: https://github.com/Vampire/setup-wsl/actions/runs/8529730666/job/23366112566

**Reply by Vampire** [reply]
Also if I copy your steps and investigate: https://github.com/Vampire/setup-wsl/actions/runs/8530166727/job/23367453214
```plain
runneradmin@fv-az1111-754 MINGW64 /d/a/setup-wsl/setup-wsl
# touch $RUNNER_TEMP/foo; $RUNNER_TEMP/wsl-shell-wrapper/wsl-bash.bat $RUNNER_TEMP/foo
/etc/profile
/root/.bashrc
```

**Reply by Vampire** [reply]
```plain
runneradmin@fv-az844-842 MINGW64 /d/a/setup-wsl/setup-wsl
# echo 'task --version' >$RUNNER_TEMP/foo; $RUNNER_TEMP/wsl-shell-wrapper/wsl-bash.bat $RUNNER_TEMP/foo                                                                                                          /etc/profile
/root/.bashrc
/d/a/_temp/foo: line 1: task: command not found
```

**Reply by Vampire** [reply]
Your problem is the first code line of `.bashrc`, which is `[ -z "$PS1" ] && return`.
This stops executing the script if `PS1` is not set which is taken as a sign that the shell is non-interactive and thus the script is stopped.

**Reply by Vampire** [reply, marked as answer]
So as I said, do not use `.bashrc` but `.profile`.
If you replace `echo '. "$HOME/.asdf/asdf.sh"' >> ~/.bashrc` by `echo '. "$HOME/.asdf/asdf.sh"' >> ~/.profile` it should work fine using login shell instead of interactive shell.

**Reply by ssbarnea** [reply]
Does it make sense to suggest this change to asdf own install instructions?

**Reply by Vampire** [reply]
I have no idea what `asdf` is and whether it makes sense.
On other distributions it might not stop executing `.bashrc`.
And if you are using an interactive shell afair `.profile` will not be sourced.
So maybe they should have both in the instructions, or a prominent warning that it might not work as intended or whatever.
But you have to discuss that with the asdf maintainers.
